sessiontxn/interface.go

```diff
@@ -15,71 +15,90 @@
 package sessiontxn
 
 import (
-	"github.com/pingcap/errors"
+	"context"
+
 	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/parser/ast"
 	"github.com/pingcap/tidb/sessionctx"
 )
 
-// TxnContextProvider provides txn context
-type TxnContextProvider interface {
-	// Initialize the provider with session context
-	Initialize(sctx sessionctx.Context) error
-	// GetTxnInfoSchema returns the information schema used by txn
-	GetTxnInfoSchema() infoschema.InfoSchema
-	// GetReadTS returns the read timestamp used by select statement (not for select ... for update)
-	GetReadTS() (uint64, error)
-	// GetForUpdateTS returns the read timestamp used by update/insert/delete or select ... for update
-	GetForUpdateTS() (uint64, error)
-}
-
-// SimpleTxnContextProvider implements TxnContextProvider
-// It is only used in refactor stage
-// TODO: remove it after refactor finished
-type SimpleTxnContextProvider struct {
-	InfoSchema         infoschema.InfoSchema
-	GetReadTSFunc      func() (uint64, error)
-	GetForUpdateTSFunc func() (uint64, error)
-}
+// EnterNewTxnType is the type to enter a new txn
+type EnterNewTxnType int
 
-// Initialize the provider with session context
-func (p *SimpleTxnContextProvider) Initialize(_ sessionctx.Context) error {
-	return nil
-}
+const (
+	// EnterNewTxnDefault means to enter a new txn. Its behavior is more straight-forward
+	// just starting a new txn right now without any scenario assumptions.
+	EnterNewTxnDefault EnterNewTxnType = iota
+	// EnterNewTxnWithBeginStmt indicates to enter a new txn when execute 'BEGIN' or 'START TRANSACTION'
+	EnterNewTxnWithBeginStmt
```

Comment:
Yes, they looks similar, but still have some differences. For example, `EnterNewTxnWithBeginStmt` can reuse the old transaction's context to avoid fetch the tso again, this optimization is strongly related to the scene and difficult to abstract. So leave it as a standalone type may be easier to understand for the current situation.